Run a single uploader service per process

zmb3 opened the pull request:
Prior to this change, each individual service (proxy, app, SSH, db, etc)
would spin up its own uploader service. If you are running multiple
Teleport services in the same process, this means you get multiple
uploaders all looking at the same directory, which can result in
duplicate upload events in the audit log:

![Screen Shot 2022-07-15 at 10 33 43 AM](https://user-images.githubusercontent.com/7527103/179267975-3185c0e6-b11d-4081-8e14-46b563c97445.png)

Additionally, desktop access has (mistakenly) failed to set up this
service, so desktop sessions would only be uploaded if you happened
to also run some other service in the same process that does spin up
the uploader.

Lastly, the uploader may not be necessary at all if sync recording modes
are used (nothing is written to disk).

Solve these issues by centralizing the uploader service so that it
runs once per process, and each Teleport service doesn't need to think
about whether or not the service should run.

Fixes #12549

zmb3 committed:
Run a single uploader service per process

Prior to this change, each individual service (proxy, app, SSH, db, etc)
would spin up its own uploader service. If you are running multiple
Teleport services in the same process, this means you get multiple
uploaders all looking at the same directory, which can result in
duplicate upload events in the audit log.

Additionally, desktop access has (mistakenly) failed to set up this
service, so desktop sessions would only be uploaded if you happened
to also run some other service in the same process that does spin up
the uploader.

Solve these issues by centralizing the uploader service so that it
runs once per process, and each Teleport service doesn't need to think
about whether or not the service should run.

Fixes #12549

zmb3 commented on lib/service/service.go:
```diff
@@ -1074,6 +1074,11 @@ func NewTeleport(cfg *Config, opts ...NewTeleportOption) (*TeleportProcess, erro
 	process.RegisterFunc("common.rotate", process.periodicSyncRotationState)
 
+	// run one upload completer per-process if we're using async recording
+	if !services.IsRecordSync(cfg.Auth.SessionRecordingConfig.GetMode()) {
```
I am leaning towards removing this check and always running the uploader service.

The concern here is if you use dynamic configuration to change the recording mode while the process is running we'll need to then start up the uploader, which feels like a pain to coordinate.

Joerger replied:
I think if we won't to shoot for the starts, ideally we would isolate all of the uploader logic to `initUploadService`, and always call it. Then, we can determine the recording strategy, and set the `StreamEmitter` a single time for all services. In the case of node sync, we simply wouldn't start the uploader service and would set the `StreamEmitter` to the `AuthClient`.

This would remove the need for the duplicated `NewStreamer` logic in every service, as well as the complicated, impossible to follow, wrapping of streamers with `AsyncStreamer`, `TeeStreamer`, etc. Not an easy task, but I would go all or nothing personally :D

zmb3 commented:
Note to reviewers: I still need to do a bunch of testing with all the various recording modes to make sure everything works correctly, but I'm pretty convinced this is the better approach and opened this in draft to get some early feedback.

Joerger reviewed (commented):
LGTM, if you remove the `IsRecordSync` check for now, with the optional follow up in the comments ;)

rosstimothy reviewed (commented):
Does leveraging `connector[0]` change the shutdown behavior at all? Is there any chance that `connector[0].Client` is closed prior to `fileuploader.shutdown` and `fileuploadcompleter.shutdown`? If so what happens to any in flight uploads?

zmb3 commented:
Moving back to draft while I validate a few more things.

zmb3 commented:
> Does leveraging `connector[0]` change the shutdown behavior at all? Is there any chance that `connector[0].Client` is closed prior to `fileuploader.shutdown` and `fileuploadcompleter.shutdown`? If so what happens to any in flight uploads?

@rosstimothy while it does look like it is possible for the connector's client to be closed before the fileuploader is shut down, this doesn't appear to be new behavior. All of this cleanup logic runs after the `TeleportExitEvent` is broadcast, and after that it's just a bunch of independent goroutines that aren't synchronized.

If you have any suggestions for how to handle this I'm all ears, otherwise I say we keep it out of scope here.

rosstimothy reviewed (commented):
>@rosstimothy while it does look like it is possible for the connector's client to be closed before the fileuploader is shut down, this doesn't appear to be new behavior. All of this cleanup logic runs after the TeleportExitEvent is broadcast, and after that it's just a bunch of independent goroutines that aren't synchronized.
> If you have any suggestions for how to handle this I'm all ears, otherwise I say we keep it out of scope here.

I just wanted to make sure that this didn't add any new/different behavior to the uploader. I'm fine with tackling the unsynchronized shutdown problem in the future.

zmb3 commented on lib/service/db.go:
```diff
@@ -149,6 +132,8 @@ func (process *TeleportProcess) initDatabaseService() (retErr error) {
 		databases = append(databases, db)
 	}
 
+	clusterName := conn.ServerIdentity.Cert.Extensions[utils.CertExtensionAuthority]
```
Nope, just got confused when resolving merge conflicts. Will update!

zmb3 committed:
address review suggestions

zmb3 committed:
Remove unused auditlog param from filesessions uploader

zmb3 committed:
Test upload completer permissions

This ensures that all builtin roles have permissions to execute
the major functions required by the upload completer service.

zmb3 committed:
Allow running auth service standalone

In this scenario, it's okay not to find a connector for the uploader
service. This is because the auth server has its own uploader service,
and if no other services are running, we don't need an additional
completer.

zmb3 commented on lib/service/service.go:
```diff
@@ -2407,12 +2399,32 @@ func (process *TeleportProcess) registerWithAuthServer(role types.SystemRole, ev
+
+	connectors := process.getConnectors()
+	var conn *Connector
+	for _, c := range connectors {
```
Nope, it shouldn't matter.

I added a test that verifies that all of the "local service" roles have adequate permissions to perform upload completer actions.

zmb3 committed:
Skip auth role in test

zmb3 committed:
Merge branch 'master' into zmb3/single-uploader-service
